grakn-graql/src/main/java/ai/grakn/graql/internal/query/runner/TinkerQueryRunner.java

```diff
@@ -90,11 +93,18 @@ public static TinkerQueryRunner create(EmbeddedGraknTx<?> tx) {
                 .flatMap(v -> v.innerVarPatterns().stream())
                 .collect(toImmutableList());
 
-        return query.admin().match().map(
-                match -> match.stream().map(answer -> QueryOperationExecutor.insertAll(varPatterns, tx, answer))
-        ).orElseGet(
-                () -> Stream.of(QueryOperationExecutor.insertAll(varPatterns, tx))
-        );
+        return query.admin().match()
+                .map(match -> runMatchInsert(match, varPatterns))
+                .orElseGet(() -> Stream.of(QueryOperationExecutor.insertAll(varPatterns, tx)));
+    }
+
+    private Stream<Answer> runMatchInsert(Match match, Collection<VarPatternAdmin> varPatterns) {
+        Set<Var> varsInMatch = match.admin().getSelectedNames();
+        Set<Var> varsInInsert = varPatterns.stream().map(VarPatternAdmin::var).collect(toImmutableSet());
+        Set<Var> projectedVars = Sets.intersection(varsInMatch, varsInInsert);
```

Comment:
aren't variables from insert enough?

Reply:
You also need the ones in the `match`. For example:
```
match $x has name "Alice";
insert (wife: $x, husband: $y) isa marriage; $y isa person, has name "Bob";
```
If we used only `insert` variables, we would project the `match` answers on `$x` and `$y`. The `match` part doesn't actually contain `$y`, so it'll throw.

Reply:
:+1: